### PR TITLE
Added comprehensive checks for flatNodes to prevent any possible pani…

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

Added comprehensive checks for flatNodes to prevent any possible panic or nil pointer:
- create flatNode without initial length, and append items as they appear (this prevent having a list with nil values for nodes that are not defined)
- added more flatNode functions, with all proper parameter checks (nil, empty values, out of range, etc)
- added comprehensive unit tests for the new changes
- added some extra flattenGeographyTopLevel test cases for getRoot or getChild returning empty responses.

This should fix the bug that the data team identified, described here: 
https://github.com/ONSdigital/dp-frontend-filter-dataset-controller/pull/282
(that PR only fixed the issue partially, hence the bug was still reproducible)

### How to review

- Make sure code changes make sense
- Make sure unit tests pass, and that the new tests make sense

### Who can review

Anyone

### CODE REVIEW

File: handlers/hierarchy.go [hunk 1]
```diff
@@ -306,35 +306,37 @@ type flatNodes struct {
...
 		Order:            val.Order,
 		NumberofChildren: val.NumberofChildren,
-	}
+	})
 }
 
 // hasOrder returns true if and only if all child items in the list have a non-nil order values
-func (n flatNodes) hasOrder() bool {
+func (n *flatNodes) hasOrder() bool {
+	if n == nil || n.list == nil {
```
Comment (ChrysmOre) [inline handlers/hierarchy.go:337]: I think it's pretty unusual to be checking a pointer receiver for a nil value. Not saying change it necessarily as it won't hurt as such, but if this is an issue that actually comes up it suggest the surrounding code is bad. It's not really a common pattern you see, it's the sort of thing that would be idiomatic to do always if it was.

Comment (DavidSubiros) [reply handlers/hierarchy.go:337]: It's not something that was required, I just started checking everything that I saw. The change that was actually needed is to check that n.list contains non-nil items (which should be fixed anyway by the changes to append values instead of writing them in particular indices).
I can remove the check for the nil pointer receiver, as you say if it is nil, there's something really wrong.

Comment (ChrysmOre) [reply handlers/hierarchy.go]: Personally I would remove them but it's completely up to you.

**Review by ChrysmOre — DISMISSED** [review]

**Review by redhug1 — APPROVED** [review]
All new changes look good to me.
Code coverage for new code is good.
